No type mapping from class java.lang.String to class java.lang.String

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-fury/issues) and found no similar issues.


### Version

fury: 0.5.0
os: linux
jdk: openjdk17
fastjson: 1.2.70

### Component(s)

Java

### Minimal reproduce step

use fury.serialize(output, obj); in dubbo spi

```java
public class TestMain {

    static ThreadSafeFury fury = Fury.builder().withLanguage(Language.JAVA)
            .requireClassRegistration(false)
            .withRefTracking(true)
            .registerGuavaTypes(false)
            .withCompatibleMode(CompatibleMode.COMPATIBLE)
            .buildThreadSafeFury();

    public static void main(String[] args) {
        DemoResponse resp = new DemoResponse();
        byte[] serialize = fury.serialize(resp);
        System.out.println("size:" + serialize.length);
    }
}


import com.alibaba.fastjson.JSONObject;
@Data
public class DemoResponse {
    private JSONObject filterJson;

    public DemoResponse() {

    }
}

```

### What did you expect to see?

serialize successfully

### What did you see instead?

```
at org.apache.fury.serializer.CodegenSerializer.loadCompatibleCodegenSerializer(CodegenSerializer.java:62)
	at org.apache.fury.resolver.ClassResolver.lambda$getObjectSerializerClass$4(ClassResolver.java:967)
	at org.apache.fury.builder.JITContext.registerSerializerJITCallback(JITContext.java:131)
	at org.apache.fury.resolver.ClassResolver.getObjectSerializerClass(ClassResolver.java:962)
	at org.apache.fury.resolver.ClassResolver.getSerializerClass(ClassResolver.java:885)
	at org.apache.fury.resolver.ClassResolver.getSerializerClass(ClassResolver.java:782)
	at org.apache.fury.resolver.ClassResolver.createSerializer(ClassResolver.java:1168)
	at org.apache.fury.resolver.ClassResolver.getOrUpdateClassInfo(ClassResolver.java:1107)
	at org.apache.fury.Fury.write(Fury.java:312)
	at org.apache.fury.Fury.serialize(Fury.java:248)
	at org.apache.fury.Fury.lambda$serialize$0(Fury.java:264)
	at org.apache.fury.Fury.serializeToStream(Fury.java:1236)
	at org.apache.fury.Fury.serialize(Fury.java:264)
```

```
Caused by: java.lang.IllegalArgumentException: No type mapping from class java.lang.String to class java.lang.String
	at org.apache.fury.reflect.TypeRef.populateTypeMappings(TypeRef.java:508)
	at org.apache.fury.reflect.TypeRef.populateTypeMappings(TypeRef.java:499)
	at org.apache.fury.reflect.TypeRef.getSubtype(TypeRef.java:458)
	at org.apache.fury.type.TypeUtils.mapOf(TypeUtils.java:513)
	at org.apache.fury.builder.CompatibleCodecBuilder.createDescriptor(CompatibleCodecBuilder.java:149)
	at org.apache.fury.builder.CompatibleCodecBuilder.lambda$buildEncodeExpression$a78febb8$4(CompatibleCodecBuilder.java:265)
	at org.apache.fury.codegen.ExpressionOptimizer.invokeGenerated(ExpressionOptimizer.java:49)
	at org.apache.fury.builder.CompatibleCodecBuilder.lambda$buildEncodeExpression$4(CompatibleCodecBuilder.java:254)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.apache.fury.builder.CompatibleCodecBuilder.buildEncodeExpression(CompatibleCodecBuilder.java:251)
	at org.apache.fury.builder.BaseObjectCodecBuilder.genCode(BaseObjectCodecBuilder.java:205)
	at org.apache.fury.codegen.CompileUnit.getCode(CompileUnit.java:53)
	at org.apache.fury.codegen.JaninoUtils.toBytecode(JaninoUtils.java:81)
	at org.apache.fury.codegen.JaninoUtils.toBytecode(JaninoUtils.java:73)
	at org.apache.fury.codegen.CodeGenerator.compile(CodeGenerator.java:145)
	at org.apache.fury.builder.CodecUtils.loadOrGenCodecClass(CodecUtils.java:110)
	at org.apache.fury.builder.CodecUtils.loadOrGenCompatibleCodecClass(CodecUtils.java:65)
	at org.apache.fury.builder.CodecUtils.loadOrGenCompatibleCodecClass(CodecUtils.java:57)
	at org.apache.fury.serializer.CodegenSerializer.loadCompatibleCodegenSerializer(CodegenSerializer.java:59)
	... 37 more
```

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

--- a/java/benchmark/src/test/java/org/apache/fury/benchmark/state/JsonTest.java
+++ b/java/benchmark/src/test/java/org/apache/fury/benchmark/state/JsonTest.java
@@ -1,0 +1,52 @@
+/*
+ * Copyright (C) 2007 The Guava Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.fury.benchmark.state;
+
+import com.alibaba.fastjson2.JSONObject;
+import org.apache.fury.Fury;
+import org.apache.fury.config.CompatibleMode;
+import org.apache.fury.config.Language;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class JsonTest {
+  public static class DemoResponse {
+    private JSONObject json;
+
+    public DemoResponse(JSONObject json) {
+      this.json = json;
+    }
+  }
+
+  @Test
+  public void testSerializeJson() {
+    // For issue: https://github.com/apache/incubator-fury/issues/1604
+    JSONObject jsonObject = new JSONObject();
+    jsonObject.put("k1", "v1");
+    jsonObject.put("k2", "v2");
+    DemoResponse resp = new DemoResponse(jsonObject);
+    Fury fury =
+        Fury.builder()
+            .withLanguage(Language.JAVA)
+            .requireClassRegistration(false)
+            .withRefTracking(true)
+            .registerGuavaTypes(false)
+            .withCompatibleMode(CompatibleMode.COMPATIBLE)
+            .build();
+    byte[] serialized = fury.serialize(resp);
+    DemoResponse o = (DemoResponse) fury.deserialize(serialized);
+    Assert.assertEquals(o.json, jsonObject);
+  }
+}

--- a/java/benchmark/src/test/java/org/apache/fury/benchmark/state/JsonTest.java
+++ b/java/benchmark/src/test/java/org/apache/fury/benchmark/state/JsonTest.java
@@ -1,15 +1,20 @@
 /*
- * Copyright (C) 2007 The Guava Authors
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
  *
- * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
- * in compliance with the License. You may obtain a copy of the License at
+ *   http://www.apache.org/licenses/LICENSE-2.0
  *
- * http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software distributed under the License
- * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
- * or implied. See the License for the specific language governing permissions and limitations under
- * the License.
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
  */
 
 package org.apache.fury.benchmark.state;

--- a/java/fury-core/src/main/java/org/apache/fury/reflect/TypeRef.java
+++ b/java/fury-core/src/main/java/org/apache/fury/reflect/TypeRef.java
@@ -498,15 +498,6 @@ public class TypeRef<T> {
       for (int i = 0; i < fromArgs.length; i++) {
         populateTypeMappings(mappings, fromArgs[i], toArgs[i]);
       }
-    } else if (supertypeWithArgsFromSubtype instanceof Class) {
-      if (toType instanceof WildcardType) {
-        return; // Okay to say Foo is <?>
-      }
-      // Can't map from a raw class to anything other than itself or a wildcard.
-      // You can't say "assuming String is Integer".
-      // And we don't support "assuming String is T"; user has to say "assuming T is String".
-      throw new IllegalArgumentException(
-          "No type mapping from " + supertypeWithArgsFromSubtype + " to " + toType);
     } else if (supertypeWithArgsFromSubtype instanceof GenericArrayType) {
       if (toType instanceof WildcardType) {
         return; // Okay to say A[] is <?>
@@ -516,7 +507,9 @@ public class TypeRef<T> {
           ((GenericArrayType) supertypeWithArgsFromSubtype).getGenericComponentType();
       populateTypeMappings(mappings, fromComponentType, componentType);
     } else {
-      throw new AssertionError("Unknown type: " + toType);
+      if (!(supertypeWithArgsFromSubtype instanceof Class)) {
+        throw new AssertionError("Unknown type: " + toType);
+      }
     }
   }
 

--- a/java/fury-core/src/test/java/org/apache/fury/reflect/TypeRefTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/reflect/TypeRefTest.java
@@ -1,15 +1,20 @@
 /*
- * Copyright (C) 2007 The Guava Authors
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
  *
- * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
- * in compliance with the License. You may obtain a copy of the License at
+ *   http://www.apache.org/licenses/LICENSE-2.0
  *
- * http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software distributed under the License
- * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
- * or implied. See the License for the specific language governing permissions and limitations under
- * the License.
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
  */
 
 package org.apache.fury.reflect;

--- a/java/fury-core/src/test/java/org/apache/fury/reflect/TypeRefTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/reflect/TypeRefTest.java
@@ -1,0 +1,37 @@
+/*
+ * Copyright (C) 2007 The Guava Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.fury.reflect;
+
+import static org.testng.Assert.*;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import org.apache.fury.type.TypeUtils;
+import org.testng.annotations.Test;
+
+public class TypeRefTest {
+  static class MapObject extends LinkedHashMap<String, Object> {}
+
+  @Test
+  public void testGetSubtype() {
+    // For issue: https://github.com/apache/incubator-fury/issues/1604
+    TypeRef<? extends Map<String, Object>> typeRef =
+        TypeUtils.mapOf(MapObject.class, String.class, Object.class);
+    assertEquals(typeRef, TypeRef.of(MapObject.class));
+    assertEquals(
+        TypeUtils.mapOf(Map.class, String.class, Object.class),
+        new TypeRef<Map<String, Object>>() {});
+  }
+}

